File: 268.Missing-Number.py
# https://leetcode.com/problems/missing-number/
#
# algorithms
# Easy (46.93%)
# Total Accepted:    230,300
# Total Submissions: 490,714

class Solution(object):
    def missingNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        length = len(nums)
        s = sum(nums)
        
        if s == (length - 1) * length / 2:
            return length
        
        return (length * (length + 1) / 2) - s
                    

File: 268.Missing-Number.java
// https://leetcode.com/problems/missing-number/
//
// algorithms
// Easy (48.10%)
// Total Accepted:    266,601
// Total Submissions: 554,267

class Solution {
    public int missingNumber(int[] nums) {
        int length = nums.length;
        
        for (int i = 0; i < length; i++) {
            while (nums[i] != i && nums[i] < length) {
                swap(nums, nums[i], i);
            }
        }
        
        for (int i = 0; i < length; i++) {
            if (nums[i] != i) {
                return i;
            }
        }
        
        return length;
    }
    
    public void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
    
}

class Solution1 {
    public int missingNumber(int[] nums) {
        int res = 0, length = nums.length;
        
        for (int i = 0; i < length; i++) {
            res ^= nums[i];
            res ^= i;
        }
        
        return res ^ length;
    }
}